server/resp/src/main/java/org/infinispan/server/resp/commands/list/blocking/BLPOP.java

```diff
@@ -0,0 +1,256 @@
+      CompletionStage<Void> addListenerStage = cache.addListenerAsync(pubSubListener, filter, converter);
+      addListenerStage.whenComplete((ignore, t) -> {
+         // If listener fails to install, complete exceptionally pubSubFuture and return
+         if (t != null) {
+            pubSubListener.completeExceptionally(t);
+         }
+         pubSubListener.setListenerAdded(true);
+         // Listener can lose events during its install, so we need to poll again
+         // and if we get values complete the listener future. In case of exception
+         // completeExceptionally
+         // Start a timer if required
+         pubSubListener.startTimer(timeout);
+         pollAllKeys(listMultimap, filterKeys).whenComplete((v, t2) -> {
+            // If second poll fails, remove listener and complete exceptionally
+            if (t2 != null) {
+               pubSubListener.completeExceptionally(t);
+            }
+            if (v != null) {
+               // If got result complete poll stage
+               pubSubListener.complete(v);
+            } else {
+               // Poll result is null. If got event from listener use it
+               if (pubSubListener.getSavedKey() != null) {
+                  pubSubListener.multimapList.pollFirst(pubSubListener.getSavedKey(), 1).thenApply(eventVal -> {
+                     pubSubListener.complete(eventVal);
+                     return null;
+                  });
+               } else {
+                  // Poll phase completed, allow listener to continue the work
+                  pubSubListener.pollComplete = true;
+               }
+            }
+         });
+      });
+      return pubSubListener.getFuture();
+
+   }
+
+   private CompletionStage<Collection<byte[]>> pollAllKeys(EmbeddedMultimapListCache<byte[], byte[]> listMultimap,
+         List<byte[]> filterKeys) {
+      var pollStage = pollKeyValue(listMultimap, filterKeys.get(0));
+      for (int i = 1; i < filterKeys.size(); ++i) {
+         var keyChannel = filterKeys.get(i);
+         pollStage = pollStage.thenCompose(
+               v -> (v == null || v.isEmpty())
+                     ? pollKeyValue(listMultimap, keyChannel)
+                     : CompletableFuture.completedFuture(v));
+      }
+      return pollStage;
+   }
+
+   CompletionStage<Collection<byte[]>> pollKeyValue(EmbeddedMultimapListCache<byte[], byte[]> mmList, byte[] key) {
+      return mmList.pollFirst(key, 1)
+            .thenApply((v) -> (v == null || v.isEmpty())
+                  ? null
+                  : Arrays.asList(key, v.iterator().next()));
+   }
+
+   @Listener(clustered = true)
+   public static class PubSubListener {
+      public EmbeddedMultimapListCache<byte[], byte[]> multimapList;
+      public AdvancedCache<Object, Object> cache;
+      ScheduledFuture<?> scheduledTimer;
+      private boolean listenerAdded;
+      private CompletableFuture<Collection<byte[]>> resultFuture;
+      public boolean pollComplete;
```

Comment:
This variable is not thread safe. We really need tests as I mentioned below to guarantee concurrent behavior as I am really worried there are a lot of gaps here.

Reply:
Only one thread changes this variable, the one that completes the poll action, I don't think it needs to be thread safe.
The only concurrent issue I can see here is addressed and commented in the code [here](https://github.com/infinispan/infinispan/blob/921467c37bb69f58f09a96a9c88eb448dd23cbfc/server/resp/src/main/java/org/infinispan/server/resp/commands/list/blocking/BLPOP.java#L228-L229).
Please be more specific about the gaps so I can work on

Reply:
It doesn't matter if only one thread changes it unfortunately. If matters if more than one thread read and/or write to it. There is no guarantee the second thread will see the update without encountering a happens before event.

I mentioned the gaps before basically concurrent insertions into the polled keys and what happens if the listener does/doesn't see the update but the second poll does or does not. Also what if listener is listening to > 1 key and if more than 1 one is updated concurrently.

Unfortunately, this command is much more complex than any other and has a high level of concurrency related to it.

Reply:
Yeah, I'm working on extracting the sync logic in a class so to make easy to write specific unit tests